### PR TITLE
Neovim: add native lsp support using lspconfig #105

### PR DESCRIPTION
**Pull request by tiagovla**
This is exactly how lspconfig expects this file to be including the description to generate their docs.

### CODE REVIEW

**Comment by tiagovla** [comment]
I have read and hereby agree to quick-lint-js' Contributor License Agreement (CLA-v1.md).

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    docs = {
+        description = [[
+https://quick-lint-js.com/
+quick-lint-js finds bugs in JavaScript programs.
+[200× faster than ESLint](https://quick-lint-js.com/benchmarks/),
+quick-lint-js gives you instant feedback as you code.
+Find bugs in your JavaScript before your finger leaves the keyboard.
+Lint any JavaScript file with no configuration.
+```sh
+npm install [-g] quick-lint-js
```
Comment (strager) [inline plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:23]: What code causes this plugin to use the quick-lint-js installed by `npm install quick-lint-js`? I see `util.find_node_modules_ancestor(fname)`, but how does that turn into running `node_modules/.bin/quick-lint-js`?

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua [hunk 2]
```diff
@@ -0,0 +1,30 @@
+local configs = require("lspconfig/configs")
```
Comment (strager) [inline plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:1]: Nit: Add copyright information. Copy-paste from other files. Also, add .lua to tools/check-file-copyrights.

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+local util = require("lspconfig/util")
+
+configs.quick_lint_js = {
+    default_config = {
+        cmd = {"quick-lint-js", "--lsp-server"},
+        filetypes = {"javascript"},
+        root_dir = function(fname)
+            return util.find_node_modules_ancestor(fname) or
+                       util.find_git_ancestor(fname) or
```
Comment (strager) [inline plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:10]: Why are we searching for Git repositories? Git seems unrelated to quick-lint-js or JavaScript projects. Is this logic standard and recommended practice for Neovim plugins?

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua (hunk 1 above)
Comment (strager) [inline plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:23]: I don't want to recommend that people install quick-lint-js through npm in order to get quick-lint-js working with their editor. Instead, people should use their system package manager (e.g. apt or Homebrew). I'd point people to the installation documentation instead: https://quick-lint-js.com/install/

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+quick-lint-js gives you instant feedback as you code.
+Find bugs in your JavaScript before your finger leaves the keyboard.
+Lint any JavaScript file with no configuration.
+```sh
+npm install [-g] quick-lint-js
+```
+]],
+        default_config = {
+            root_dir = [[root_pattern('package.json', 'node_modules', '.git') or dirname]]
+        }
```
Comment (strager) [inline plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:28]: Why is this sorta repeated twice? Why does `root_dir` need documentation? Is this normal? I'm confused. =S (Maybe I should read the documentation.)

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua (hunk 3 above)
Comment (tiagovla) [reply plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:16]: Those are ways to choose the ``root_dir`` which will spawn a server instance. Since most configs in lspconfig are like that, I just kept it the same way.

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua (hunk 4 above)
Comment (tiagovla) [reply plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua:34]: That will make lspconfig documentation look like [this](https://github.com/neovim/nvim-lspconfig/blob/master/CONFIG.md#rome).
```lua
require'lspconfig'.quick_lint_js.setup{}

  Commands:
  
  Default Values:
    cmd = { "quick-lint-js", "--lsp-server" }
    filetypes = { "javascript" }
    root_dir = root_pattern('package.json', 'node_modules', '.git') or dirname
```
Instead of showing those util functions.

File: plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua (hunk 1 above)
Comment (tiagovla) [reply plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua]: That's the  ``cmd = {"quick-lint-js", "--lsp-server"},`` which expects **quick-lint-js** to be in ``$PATH``. It's possible to make it look somewhere and replace it instead, do you have a helper function to find its path?

Comment (strager) [reply plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua]: > That's the `cmd = {"quick-lint-js", "--lsp-server"},` which expects **quick-lint-js** to be in `$PATH`.

Right. So why do the instructions recommend running `npm install quick-lint-js` if that won't put `quick-lint-js` in `$PATH`?

> It's possible to make it look somewhere and replace it instead, do you have a helper function to find its path?

I think it'd be `node_modules/.bin/quick-lint-js`. However, this might be a security risk.

What does ESLint's integration do?

Comment (tiagovla) [reply plugin/vim/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua]: That's the default configuration. You either put it in ``$PATH`` and use it like this:
```lua
require("lspconfig").quick_lint_js.setup{}
```
or you overwrite the default config in your vimrc/init.lua like this:
```lua
require("lspconfig").quick_lint_js.setup{ cmd = "/somewhere/node_modules/.bin/quick-lint-js" }
```
There are some plugins like [lspinstall](https://github.com/kabouzeid/nvim-lspinstall) or the one I was making [lspconfigplus](https://github.com/tiagovla/lspconfigplus.nvim) that manages installation and update of lsp servers. They pretty much install it somewhere and overwrite that ``cmd`` key to point to the executable, seamlessly.

ESLint is not listed [there](https://github.com/neovim/nvim-lspconfig/tree/master/lua/lspconfig). All of them consider the executable to be in ``$PATH``.

**Comment by tiagovla** [comment]
> Blocking: Include a README file describing how to install the plugin manually. (If the existing README applies, please update it to mention this stuff as alternative to ALE.)
> 
I added lspconfig and Packer to the docs.
> Does the manual installation documentation on the website work for this plugin? https://quick-lint-js.com/install/#manual-vim
> 
If ale is working, in theory it should, I will test it somewhere in the near future.
> 
> We should make sure that this plugin works out of the box as much as possible for the Debian package. See dist/debian/README.md for instructions. If you don't have a Debian machine, I can help test or debug.

I use arch but I have some ubuntu servers. I might test it there.

**Comment by strager** [comment]
I'm a Neovim noob. Help me out. Here's what I did:

1. Download the [Neovim nightly AppImage](https://github.com/neovim/neovim/releases/tag/nightly). I installed version v0.5.0-dev+1339-g5d9c56012 for Linux.
2. Create ~/.config/nvim/init.lua with the following content: `require("lspconfig").quick_lint_js.setup{}`
3. Run Neovim.

Here's what I see when I open Neovim:

```
E5113: Error while calling lua chunk: /home/strager/.config/nvim/init.lua:1: module 'lspconfig' not 
found:
        no field package.preload['lspconfig']
        no file './lspconfig.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/luajit-2.1.0-beta3/lspconfig.lua'
        no file '/usr/local/share/lua/5.1/lspconfig.lua'
        no file '/usr/local/share/lua/5.1/lspconfig/init.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/lspconfig.lua'
        no file '/home/runner/work/neovim/neovim/.deps/usr/share/lua/5.1/lspconfig/init.lua'
        no file './lspconfig.so'
        no file '/usr/local/lib/lua/5.1/lspconfig.so'
        no file '/home/runner/work/neovim/neovim/.deps/usr/lib/lua/5.1/lspconfig.so'
        no file '/usr/local/lib/lua/5.1/loadall.so'
Press ENTER or type command to continue
```

Apparently I need to install https://github.com/neovim/nvim-lspconfig. (I thought it was built into Neovim.) I did so (`git clone https://github.com/neovim/nvim-lspconfig ~/.config/nvim/start/nvim-lspconfig`), then opened Neovim again. This is what I see now:

```
E5113: Error while calling lua chunk: /home/strager/.config/nvim/init.lua:1: attempt to index field 
'quick_lint_js' (a nil value)
Press ENTER or type command to continue
```

I installed quick-lint-js via Debian (`dist/debian/`), but apparently the Debian package didn't install the Neovim plugin in the right spot.

I installed quick-lint-js' Neovim plugin by running the following command, but it seemed to not change anything:
```
$ # My quick-lint-js Git checkout is in ~/Projects/quicklint-js.
$ ln -s ~/Projects/quicklint-js/plugin/vim/quick-lint-js.vim ~/.config/nvim/start/
```

~/.config/nvim/start/quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua exists and contains your script. `runtimepath` includes `quick-lint-js.vim`:

```
runtimepath=~/.config/nvim,~/.config/nvim/start/quick-lint-js.vim,~/.config/nvim/start/nvim-lspcon
fig,/etc/xdg/xdg-cinnamon/nvim,/etc/xdg/nvim,~/.local/share/nvim/site,/usr/share/cinnamon/nvim/site,
/usr/share/gnome/nvim/site,~/.local/share/flatpak/exports/share/nvim/site,/var/lib/flatpak/exports/s
hare/nvim/site,/usr/local/share/nvim/site,/usr/share/nvim/site,/tmp/.mount_nvim.aEv35Xu/usr/share/nv
im/runtime,/tmp/.mount_nvim.aEv35Xu/usr/share/nvim/runtime/pack/dist/opt/matchit,/tmp/.mount_nvim.aE
v35Xu/usr/lib/nvim,/usr/share/nvim/site/after,/usr/local/share/nvim/site/after,/var/lib/flatpak/expo
rts/share/nvim/site/after,~/.local/share/flatpak/exports/share/nvim/site/after,/usr/share/gnome/nvim
/site/after,/usr/share/cinnamon/nvim/site/after,~/.local/share/nvim/site/after,/etc/xdg/nvim/after,/
etc/xdg/xdg-cinnamon/nvim/after,~/.config/nvim/after
```

Am I missing an installation step?

## Comments

* I thought the LSP stuff was built into Neovim (v0.5), but apparently we need to install a separate plugin. Can we make our plugin work out of the box without installing a separate plugin too?
* Documentation is lacking. It doesn't mention adding `require("lspconfig").quick_lint_js.setup{}` to ~/.config/nvim/init.lua. (Do the Packer instructions do this automatically? I did a manual install, since your PR hasn't merged yet.)
* The Debian package (quick-lint-js-vim) should put files in the right spot such that Neovim can find them without extra hassle.

**Comment by tiagovla** [comment]
### Here's how I do it:
1 - Install neovim:
```bash
git clone https://github.com/neovim/neovim.git neovim
cd neovim
make CMAKE_BUILD_TYPE=RelWithDebInfo
sudo make install
```
2 - Install the package manager Packer:
```bash
git clone https://github.com/wbthomason/packer.nvim\ ~/.local/share/nvim/site/pack/packer/start/packer.nvim
```
3 - Edit your **init.lua** inside ``$HOME/.config/nvim``:
```lua
require("packer").startup(function()
    use "wbthomason/packer.nvim"
    use "neovim/nvim-lspconfig"
    use {"$HOME/github/quick-lint-js/plugin/vim/quick-lint-js.vim"}
end)

require("lspconfig/quick_lint_js")
require("lspconfig").quick_lint_js.setup {}
```
4 - Run neovim, ignore require erros the first time (packages not installed yet), run:
```vimscript
:PackerInstall
```
5 - Restart neovim

### Extra:
This ``require("lspconfig/quick_lint_js")`` is now needed because they recently (it was working a couple weeks ago without it) decided to save a few microseconds using [this](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig.lua#L89-L91) instead a normal **require**, so now it needs the full path of the file instead of the relative path which would find the config inside **quick-lint-js.vim**.

If quick-lint-js is not in ``$PATH``, change the setup to something like this:
```lua
require("lspconfig").quick_lint_js.setup {
    cmd = {
        "/home/tiagovla/.local/share/nvim/servers/quick-lint-js/node_modules/.bin/quick-lint-js",
        "--lsp-server"
    }
}
```
If you want to debug the lsp server, add this line to your init.lua:
```lua
vim.lsp.set_log_level("debug")
--logs inside ~/.cache./nvim/lsp.log
```
If you want to check if the server is working, use ``:LspInfo``.

Did not test, but it should work with:
```lua
use {"strager/quick-lint-js", rtp = "plugin/vim/quick-lint-js.vim"}
```
### Comments:

>     * I thought the LSP stuff was built into Neovim (v0.5), but apparently we need to install a separate plugin. Can we make our plugin work out of the box without installing a separate plugin too?
> 

Lsp is builtin, you don't need a plugin for that. **Lspconfig** is an optional maintained configuration plugin, it makes it easier to setup a language server and it's what everyone is using. It's possible to make a custom one.

>     * Documentation is lacking. It doesn't mention adding `require("lspconfig").quick_lint_js.setup{}` to ~/.config/nvim/init.lua. (Do the Packer instructions do this automatically? I did a manual install, since your PR hasn't merged yet.)
> 
I thought I had put it there, I guess I forgot about it.

>     * The Debian package (quick-lint-js-vim) should put files in the right spot such that Neovim can find them without extra hassle.
>
This should work like ALE works. I cannot test it though.

**Comment by tiagovla** [comment]
Since you manually installed and have the following:

1. ``~/.config/nvim/start/nvim-lspconfig``
2. ``~/.config/nvim/start/quick-lint-js.vim``

This should be enough to get it working in your init.lua:
```lua
require("lspconfig/quick_lint_js") -- now required because lspconfig made a stupid optimization
require("lspconfig").quick_lint_js.setup {}
```
Now it should not give you this error anymore:
> ```
> E5113: Error while calling lua chunk: /home/strager/.config/nvim/init.lua:1: attempt to index field 
> 'quick_lint_js' (a nil value)
> Press ENTER or type command to continue
> ```

**Comment by strager** [comment]
Do you know why lines with multiple errors get more and more ■s? See this screenshot:
Screenshot from 2021-05-21 20-23-07

File: plugin/vim/README.md [hunk 5]
```diff
@@ -79,7 +79,12 @@ This option requires [Neovim][] 0.5 or newer and [Packer][]
 Add the following line to your `init.lua` file between
 `require('packer').startup(function()` and `end)`:
 
-    use('strager/quick-lint-js', rtp = 'plugin/vim/quick-lint-js.vim')
+    use('strager/quick-lint-js', rtp = 'plugin/vim/quick-lint-js.vim', requires = {'neovim/nvim-lspconfig'})
+
+Then, add the following lines to your `init.lua` file:
+
+    require('lspconfig/quick_lint_js')
+    require('lspconfig').quick_lint_js.setup {}
```
Comment (strager) [inline plugin/vim/README.md:87]: I think these instructions should be in a separate section for configuring. Reasons why:

* These instructions are also needed if you install using Vim-Plugged or another plugin manager.
* These instructions are only relevant when using Neovim + nvim-lspconfig.

(I've been meaning to add a section for configuring the ALE plugin (#175).)

**Comment by strager** [comment]
If you rebase atop commit 14e2e05704502d50f3577ba232f9bca6e5fe2a93 (landed on master), then apply the following patch, then our plugin works assuming you have nvim-lspconfig installed and you have the quick-lint-js plugin configured:
```diff
diff --git a/plugin/vim/CMakeLists.txt b/plugin/vim/CMakeLists.txt
index 4e344342..93b433dd 100644
--- a/plugin/vim/CMakeLists.txt
+++ b/plugin/vim/CMakeLists.txt
@@ -21,6 +21,10 @@ install(
   FILES quick-lint-js.vim/ale_linters/javascript/quick_lint_js.vim
   DESTINATION "${QUICK_LINT_JS_INSTALL_NEOVIM_DIR}/ale_linters/javascript"
 )
+install(
+  FILES quick-lint-js.vim/lua/lspconfig/quick_lint_js.lua
+  DESTINATION "${QUICK_LINT_JS_INSTALL_NEOVIM_DIR}/lua/lspconfig"
+)
 
 # quick-lint-js finds bugs in JavaScript programs.
 # Copyright (C) 2020  Matthew Glazar
```

**Comment by strager** [comment]
I rebased and tweaked this PR in #458. (I'll still give you credit, 'cus you got the basics working and explained to me how things work.) Feel free to review #458.

**Comment by tiagovla** [comment]
> I rebased and tweaked this PR in #458. (I'll still give you credit, 'cus you got the basics working and explained to me how things work.) Feel free to review #458.

Sorry, I had a lot going on at my university these past months. And #458 looks great. We might also need to test the auto completion feature with [nvim-cmp](https://github.com/hrsh7th/nvim-cmp).
